Author manuscript - allow for errors and record appropriate notes

emanuil-tolev opened the pull request:
This is tested - the errors do appear correctly. Tested by pointing europepmc.org and ebi.ac.uk to 127.0.0.1 and setting my local nginx to return 500 for those server names.

Here is a file showing what it looks like with EPMC _enabled_ (so the usual case 99% of the time) - so you can see there's no ill effects there.

[author_manuscript_test_results.csv.zip](https://github.com/CottageLabs/api/files/432192/author_manuscript_test_results.csv.zip)

emanuil-tolev committed:
Handle HTTP errors related to author manuscript or full text XML

markmacgillivray commented:
@emanuil-tolev  I don't think this is a good solution. You have removed the lower level catch when eupmc fulltext is attempted to be accessed - why? If it cannot return what you expect, would it not be better to have the upstream functions check to see if they get what they expect, and update the eupmc fulltext function to return some useful info about the error if you need it? Also for some reason you seem to have totally changed the way the API URL route returns the XML that it finds from europepmc - it previously served the xml back to you so you could see it directly from our API, but you have taken that out altogether, as far as I can see. Why?

emanuil-tolev commented:
> Also for some reason you seem to have totally changed the way the API URL route returns the XML that it finds from europepmc - it previously served the xml back to you so you could see it directly from our API, but you have taken that out altogether, as far as I can see. Why?

Not sure - it should still be returning everything. There was some unreachable code (after a return statement) that's been removed, but the actual return should be exactly the same.

> You have removed the lower level catch when eupmc fulltext is attempted to be accessed - why? If it cannot return what you expect, would it not be better to have the upstream functions check to see if they get what they expect, and update the eupmc fulltext function to return some useful info about the error if you need it?

I was mostly trying to think (or rather code) through the problem, and I'm used to using exceptions to pass info around. I very much expected this might not be a great solution, and looking at it I agree that duplicating the exception handler in multiple places is probably not good - the exception should be handled at a lower level (as it was before). So I'll do a slightly different approach where `fulltextXML` handles possible exceptions, but returns different values to allow the main service to record its compliance notes properly.

markmacgillivray commented:
@emanuil-tolev for the europepmc fulltext route, yeah I see what you are saying about the return statement, but in that case it is the return statement that is wrong and should be removed. I have tested this on dev and it works properly when that return statement is taken out, I do not know when it got added, but it is the error. It probably would not have been noticed because I only use the route for convenience, but it would still be better to have it working than not. But I have changed my edit back in dev to avoid a merge conflict, so fix it up when you make your error catching change.

Re. the error catching change, yeah it would be better to catch it where it occurs I think and then pass more useful info up if necessary. Or just the fact that it cannot return useful data could be enough. This is how we altered other lower level routes so that lantern could record provenance info, but without having to put catches all over the place.

Make the changes then I will review again, thanks.

emanuil-tolev commented:
> Or just the fact that it cannot return useful data could be enough.

Indeed, the problem I had in particular is that 404 != 500, and the two error conditions are significant on their own.

emanuil-tolev committed:
catch EPMC Fulltext fetch errors in the fetch function and return an envelope with the error

emanuil-tolev committed:
remove obsolete comment

emanuil-tolev committed:
fix EuropePMC dev use route

emanuil-tolev commented:
Okay, this is refactored now!

> @emanuil-tolev for the europepmc fulltext route, yeah I see what you are saying about the return statement, but in that case it is the return statement that is wrong and should be removed. I have tested this on dev and it works properly when that return statement is taken out, I do not know when it got added, but it is the error. It probably would not have been noticed because I only use the route for convenience, but it would still be better to have it working than not. But I have changed my edit back in dev to avoid a merge conflict, so fix it up when you make your error catching change.

I've done this now - removed the faulty return and left in the bespoke response writing code (where it outputs headers and content "manually"). Tested the route and works for me. Also tested the two error handling parts (404 and 500, the 500 by faking that www.ebi.ac.uk returns 500s), all works.

> You have removed the lower level catch when eupmc fulltext is attempted to be accessed - why? If it cannot return what you expect, would it not be better to have the upstream functions check to see if they get what they expect, and update the eupmc fulltext function to return some useful info about the error if you need it?

I've redone this so exceptions are no longer used to pass information upstream. I didn't want to overload a single primitive return value with so many different meanings:
- when boolean false (or alternatively undefined) then we didn't get the fulltext
- when it's a certain string, then it's a 404 error from EPMC that we need to add a note about
- when it's another certain string then it's a 500 error and we need to add a different note
- when it's any other string it's the fulltext

So instead of doing that ^, `CLapi.internals.use.europepmc.fulltextXML` now returns an object which is used as an envelope. `.fulltext` gives you the fulltext string, `.error` gives you error information, if you care about it. This keeps all the many `if (ft)` conditionals in the code as they are - all I had to do was make sure `ft` was the string fulltext, not the envelope obj.

emanuil-tolev committed:
try a bit harder for AAM if the function is used a certain way by passing 2 args in

emanuil-tolev commented:
One more minor change added now - make the AAM detect function try a bit harder in the rare but possible case that it gets both a `fulltext` and a `pmcid` passed in. The current code does not use the function in this manner, it actually never uses the `fulltext` argument. So this is more of a protection against an easy to introduce bug (just call the f() with all args) than a fix for a problem.

markmacgillivray commented:
Why write special error codes in strings that then have to be checked later to write more strings based on those strings, where only those strings can be understood by the code that knows to look for those strings? Why not just pass back the error codes themselves?

emanuil-tolev committed:
pass and interpret the HTTP status codes directly when there is a problem fetching FT XML

emanuil-tolev commented:
> Why write special error codes in strings that then have to be checked later to write more strings based on those strings, where only those strings can be understood by the code that knows to look for those strings? Why not just pass back the error codes themselves?

Good point, I've changed it now. We kind of do that elsewhere, but in those places we interpret different kinds of _success_ differently, whereas here we have the same errors needing to be interpreted by the calling code.
